### ISSUE TITLE
Simple query taking forever until a request timeout error

### ISSUE THREAD

**Issue by cscetbon** [arangodb/arangodb#4651]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [X] 3.3

Mode:
- [X] Cluster

Storage-Engine:
- [X] rocksdb

On this operating system:
- [X] Linux 
  - [X ] docker - official docker library
- [X] MacOS, version:

### this is an AQL-related issue:
[X] I'm using graph features

I'm issuing AQL via:
- [X] web interface with this browser:     running on this OS: OS X
- [X] arangosh

I've run `db._explain("<my aql query>")` and it didn't shed more light on this.
The AQL query in question is:

```
FOR g IN groups    
  FILTER g._key == "g1"
  let k = (
    INSERT 
    {
        _key : 'u1_g1', 
        cuid : 'u1' 
    } IN demands 
    RETURN NEW._key
  )
  RETURN k
```

The issue can be reproduced using this dataset:

```
#!arangosh --javascript.execute
db._createDatabase("db1");
db._useDatabase("db1");

let collectionProperties =  { "numberOfShards" : 5 , "waitForSync" : true, "replicationFactor" : 3 }
const groups = db._createDocumentCollection("groups", collectionProperties);

[
{"_key":"g1","mail":"mykey@ddd.com"}
].forEach(function (group) {
    groups.save(group);
});

const demands = db._createDocumentCollection("demands", collectionProperties);
```

These are the steps to reproduce:
1) open the browser on http://127.0.0.1:8529
2) log in as root
3) use database `db1` 
4) click to on queries
5) execute the query provided earlier

Here is the query plan :

```
Query string:
 FOR g IN groups    
   FILTER g._key == "g1"
   let k = (
     INSERT 
     {
         _key : 'u1_g1', 
         cuid : 'u1' 
     } IN demands 
     RETURN NEW._key
   )
   RETURN k

Execution plan:
 Id   NodeType          Site       Est.   Comment
  1   SingletonNode     DBS           1   * ROOT
 12   IndexNode         DBS           1     - FOR g IN groups   /* primary index scan */
 19   RemoteNode        COOR          1       - REMOTE
 20   GatherNode        COOR          1       - GATHER 
 10   SubqueryNode      COOR          1       - LET k = ...   /* subquery */
  5   SingletonNode     DBS           1         * ROOT
  6   CalculationNode   DBS           1           - LET #7 = { "_key" : "u1_g1", "cuid" : "u1" }   /* json expression */   /* const assignment */
 13   DistributeNode    DBS           1           - DISTRIBUTE
 14   RemoteNode        COOR          1           - REMOTE
  7   InsertNode        COOR          1           - INSERT #7 IN demands
  8   CalculationNode   COOR          1           - LET #9 = $NEW.`_key`   /* attribute expression */
 11   ReturnNode        COOR          1       - RETURN k

Indexes used:
 By   Type      Collection   Unique   Sparse   Selectivity   Fields       Ranges
 12   primary   groups       true     false            n/a   [ `_key` ]   (g.`_key` == "g1")

Optimization rules applied:
 Id   RuleName
  1   remove-data-modification-out-variables
  2   use-indexes
  3   remove-filter-covered-by-index
  4   remove-unnecessary-calculations-2
  5   distribute-in-cluster
  6   scatter-in-cluster
  7   distribute-filtercalc-to-cluster
  8   remove-unnecessary-remote-scatter

Write query options:
 Option                   Value
 ignoreErrors             false
 waitForSync              false
 nullMeansRemove          false
 mergeObjects             true
 ignoreDocumentNotFound   false
 readCompleteInput        false
 useIsRestore             false
 consultAqlWriteFilter    false
```
The following problem occurs: 
It runs for a very long time consuming memory and cpu resource until a request timeout is returned. It runs well in single instance mode though. One thing to note is that if I get rid of the let statement and just return the value, it runs as expected. This query is longer than you can see, that's why I need to assign the return value to a variable. But having narrowed it down to that piece of code I think it should help.
 
Instead I would be expecting: 
It should quickly return the same value as in single mode.

**Comment by dothebart** [comment]
From the explain, you can see that you have two sets of Scatter/gather nodes, which transfer data from/to the DB-Servers and the Coordinator. 

It also seems that you don't do any filtering, so the amount of data you transfer in these transitions is really huge. 

You should try to reduce the size of the data or the number of documents sent back and forth to improve the performance of the query.

**Comment by jsteemann** [comment]
@dothebart : It is actually a bug, and should be fixed via the above PRs.

**Comment by cscetbon** [comment]
Thanks @jsteemann. Are there nightly builds that I could test to see if it's fixed without having to wait for the next release ?

**Comment by jsteemann** [comment]
Yes, there are nightly builds (available [here](https://www.arangodb.com/nightly-builds/) as Docker containers).
However, the fix I made is currently available in the PR only, but is not contained in our "stable" 3.3 branch. When it gets reviewed and approved, it will be merged back into 3.3 and then appear in the next nightly build and/or proper release.

**Comment by cscetbon** [comment]
Yeah understood, thanks Jan
